[BE][CI] bump `ruff` to 0.9.2: multiline `assert` statements

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145606
* __->__ #144546

Reference: https://docs.astral.sh/ruff/formatter/black/#assert-statements

> Unlike Black, Ruff prefers breaking the message over breaking the assertion, similar to how both Ruff and Black prefer breaking the assignment value over breaking the assignment target:
>
> ```python
> # Input
> assert (
>     len(policy_types) >= priority + num_duplicates
> ), f"This tests needs at least {priority+num_duplicates} many types."
>
>
> # Black
> assert (
>     len(policy_types) >= priority + num_duplicates
> ), f"This tests needs at least {priority+num_duplicates} many types."
>
> # Ruff
> assert len(policy_types) >= priority + num_duplicates, (
>     f"This tests needs at least {priority + num_duplicates} many types."
> )
> ```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

XuehaiPan committed:
Update

[ghstack-poisoned]

Skylion007 commented:
Doesn't black need to be updated too in lintrunner? I think most of the changes in this ruff are to align with black's 2024 style?

Skylion007 commented:
> > Doesn't black need to be updated too in lintrunner?
> 
> No. A file is either formatted by `black` (23.12.1) or `ruff format`. Not both.
> 
> https://github.com/pytorch/pytorch/blob/a222029f4e0233a263eb61e5a37f2cd01e1f3c66/tools/linter/adapters/pyfmt_linter.py#L184-L187
> 
> And `black` is going to be removed anyway.

True, but while we support as an option, we need to update black so it can support the 2025 Black stable style

XuehaiPan commented:
> True, but while we support as an option, we need to update black so it can support the 2025 Black stable style

Bumping the code style needs to be done in a single PR and cannot be split into multiple PRs. That could touch all Python files in the codebase (wrapping for assertion statements and quote styles in nested f-strings). That is impossible to review and get the patch to be merged.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
> > True, but while we support as an option, we need to update black so it can support the 2025 Black stable style
> 
> Bumping the code style needs to be done in a single PR and cannot be split into multiple PRs. That could touch all Python files in the codebase (wrapping for assertion statements and quote styles in nested f-strings). That is impossible to review and get the patch to be merged.

Let me elaborate more on this.

- bump `ruff 0.8.4 -> 0.9.0`: touches 128 files.
- bump `ruff 0.8.4 -> 0.9.0` & `black 23.12.1 -> 24.10.0`: touches 128 + 276 files.

XuehaiPan committed:
Update

[ghstack-poisoned]

malfet commented on .github/scripts/label_utils.py:
```diff
@@ -63,9 +63,9 @@ def gh_get_labels(org: str, repo: str) -> List[str]:
-    assert (
-        last_page > 0
-    ), "Error reading header info to determine total number of pages of labels"
+    assert last_page > 0, (
+        "Error reading header info to determine total number of pages of labels"
+    )
```
Can you post a reference to a document of sorts explaining why one is better than another?

XuehaiPan replied:
This is documented in astral-sh/ruff#15127.

This is one of the styles in `ruff format` that differs from `black`: https://docs.astral.sh/ruff/formatter/black/#assert-statements

> [`assert` statements](https://docs.astral.sh/ruff/formatter/black/#assert-statements)
> Unlike Black, Ruff prefers breaking the message over breaking the assertion, similar to how both Ruff and Black prefer breaking the assignment value over breaking the assignment target:
> 
> ```python
> # Input
> assert (
>     len(policy_types) >= priority + num_duplicates
> ), f"This tests needs at least {priority+num_duplicates} many types."
> 
> 
> # Black
> assert (
>     len(policy_types) >= priority + num_duplicates
> ), f"This tests needs at least {priority+num_duplicates} many types."
> 
> # Ruff
> assert len(policy_types) >= priority + num_duplicates, (
>     f"This tests needs at least {priority + num_duplicates} many types."
> )
> ```

malfet replied:
Is there a way to force ruff to stick to the black convention? Or it's a either or situation now?

XuehaiPan replied:
> Is there a way to force ruff to stick to the black convention?

No. `ruff format` does not provide an option. Also, we are using "black 2023 style", and we are now in 2025.

Currently, we are using two formatters for different sets of files. The goal is to migrate the formatter for all Python files to `ruff format` and retire `black`. We can add this PR to git blame ignore after it is merged.

malfet reviewed (changes requested):
Can you please split it in two(or may be 3 PRs):
 - One that fixes `f'{x}'` to `f"{x}"` (which is reasonable rules we've adopted a while back)
 - One with multiline asserts (referencing the rule that explains one one is better than another, or perhaps can we suppress this rule from ruff?)
 - One that finally updates it to 0.9.0?

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
> One that fixes f'{x}' to f"{x}" (which is reasonable rules we've adopted a while back)

I split this into a separate PR:

- #144569

------

> One with multiline asserts (referencing the rule that explains one one is better than another, or perhaps can we suppress this rule from ruff?)
> One that finally updates it to 0.9.0?

These cannot be split into two RPs. I keep the, in this PR.

I have updated the PR description referring to the ruff formatter docs. The `assert` statement style is a ruff-specific derivation that differs from the black style. There is no option to disable this.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

Skylion007 commented:
Why not go directly to ruff 0.9.1?

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
cc @malfet and @Skylion007 for code review.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
FYI, ruff 0.9.8 was just released.

malfet reviewed (approved):
Sure, but if they'll change formatting again in say ruff-0.9.5 then I'll be against another reformatting

malfet commented:
@pytorchbot merge -f "Lint is green, hope it will not coderot"
